torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+  // Constant terms are stored and handled separately from other inputs. For
+  // example, if I have a + b + 3 + c + 5, then the constant term will be 8, and
+  // a, b, c are all inputs.
+  bool hasConstantTerm() const {
+    return attributes().size() == 2;
+  }
+
+  Val* getConstantTerm() const {
+    return attributeVal(1);
+  }
+
+  template <typename T>
+  T getConstantTermValue() const {
+    return *getConstantTerm()->as<Scalar<T>>()->value();
+  }
+
+  // Get a vector of inputs, sorted as the order given by `variables`. Note that
+  // the sorting key is the rightmost variable that an input depends on. For
+  // example, if I have two inputs.
+  // v1 = a * c
+  // v2 = b
+  // and variables is [a, b, c], then v2 < v1 because the rightmost depending
+  // variable of v2 is b, and the rightmost depending variable of v1 is c,
+  // and b < c. So in this example, this function will return [v2, v1].
+  // Tensors are always considered as variables and they are always considered
+  // as the rightmost.
+  std::vector<Val*> sortedInputs(const std::list<ValInfo>& variables) {
+    std::unordered_set<Val*> variables_set;
+    variables_set.reserve(variables.size());
+    for (auto v : variables) {
+      variables_set.emplace(v.variable);
+    }
+    std::vector<Val*> sorted_inputs(inputs().begin(), inputs().end());
+    auto compare = [&](Val* v1, Val* v2) {
+      bool v1_lefter_than_v2 = false;
+      auto deps1 = getSubexprDependency(v1, variables_set);
+      auto deps2 = getSubexprDependency(v2, variables_set);
+      if (hasTensor(deps2)) {
+        return true;
+      }
+      if (hasTensor(deps1)) {
+        return false;
+      }
+      for (auto v : variables) {
+        if (deps1.count(v.variable) > 0 && deps2.count(v.variable) == 0) {
+          v1_lefter_than_v2 = false;
+        } else if (
+            deps2.count(v.variable) > 0 && deps1.count(v.variable) == 0) {
+          v1_lefter_than_v2 = true;
+        }
+      }
+      return v1_lefter_than_v2;
+    };
+    std::sort(sorted_inputs.begin(), sorted_inputs.end(), compare);
+    return sorted_inputs;
+  }
+
+ protected:
+  // Add a new value as an input of this expression. If `value` is a constant
+  // scalar, then evalate this constant and merge it with the constant term.
+  // Otherwise, add it as an input.
+  void addInput(Val* value) {
+    TORCH_CHECK(
+        hasSimilarType(dtype(), *value->getDataType()),
+        "Input types should be similar, but got: ",
+        dtype(),
+        ", and ",
+        *value->getDataType());
+    value = foldConstants(value);
+    if (hasValue(value)) {
+      updateConstantTermWith(value);
+    } else {
+      Expr::addInput(value);
+    }
+  }
+
+  // Update the constant term. For example, if getOpType() == "+", and the
+  // current constant term is 8, then updateConstantTermWith(7) will change the
+  // constant term to 15.
+  void updateConstantTermWith(Val* value) {
+    if (!hasConstantTerm()) {
+      addAttribute(value);
```

Comment:
I understand why you're doing this but it makes me nervous to have `addAttribute` in a function not called on construction. I'm just worried someone could put another `addAttribute` somewhere in this Expr and get strange results. Just a nit though.

Reply:
`updateConstantTermWith` is `protected`, so the only way for this to happen is via subclass.

Reply:
I guess the concern is we generally follow the pattern that `addInput/addOutput/addAttribute` are only used inside constructors so that those inputs/outputs/attributes are immutable. There's no strict enforcing contract, but that's been a general guideline. Having those functions used in non constructors might give make someone think it's okay to modify those attributes outside constructors, which we would want to avoid. I guess we could have some enforcement with a passkey-like technique, but not sure if it's worth doing so.